hi-ml/testhiml/testhiml/test_regression_tests.py

```diff
@@ -0,0 +1,239 @@
+    actual = tmp_path / "actual.does_not_matter"
+    expected.write_text(
+        """foo,bar
+1.0,10.0"""
+    )
+    actual.write_text(
+        """foo,bar
+1.0001,10.001"""
+    )
+    assert compare_files(expected=expected, actual=actual, csv_relative_tolerance=1e-2) == ""
+    assert compare_files(expected=expected, actual=actual, csv_relative_tolerance=1e-3) == ""
+    assert compare_files(expected=expected, actual=actual, csv_relative_tolerance=2e-4) == ""
+    assert compare_files(expected=expected, actual=actual, csv_relative_tolerance=9e-5) == CONTENTS_MISMATCH
+
+
+def test_compare_files_empty_csv(tmp_path: Path) -> None:
+    """
+    If either of the two CSV files is empty, it should not raise an error, but exit gracefully.
+    """
+    expected = tmp_path / "expected.csv"
+    actual = tmp_path / "actual.csv"
+    valid_csv = """foo,bar
+    1.0,10.0"""
+    empty_csv = ""
+    for expected_contents, actual_contents in [(empty_csv, empty_csv), (valid_csv, empty_csv), (empty_csv, valid_csv)]:
+        expected.write_text(expected_contents)
+        actual.write_text(actual_contents)
+        assert compare_files(expected=expected, actual=actual) == FILE_FORMAT_ERROR
+    expected.write_text(valid_csv)
+    actual.write_text(valid_csv)
+    assert compare_files(expected=expected, actual=actual) == ""
+
+
+@pytest.mark.parametrize("file_extension", [".png", ".whatever"])
+def test_compare_files_binary(tmp_path: Path, file_extension: str) -> None:
+    """
+    Checks the comparison of files that are not recognized as text files, for example images.
+    :param test_output_dirs:
+    :param file_extension: The extension of the file to create.
+    """
+    expected = tmp_path / f"expected{file_extension}"
+    actual = tmp_path / "actual.does_not_matter"
+    data1 = bytes([1, 2, 3])
+    data2 = bytes([4, 5, 6])
+    expected.write_bytes(data1)
+    actual.write_bytes(data1)
+    assert compare_files(expected=expected, actual=actual) == ""
+    actual.write_bytes(data2)
+    assert compare_files(expected=expected, actual=actual) == CONTENTS_MISMATCH
+
+
+def test_compare_folder(tmp_path: Path) -> None:
+    """
+    Test the comparison of folders that we use for regression tests.
+    """
+    # Create a test of expected and actual files on the fly.
+    expected = tmp_path / "expected"
+    actual = tmp_path / "actual"
+    matching = "matching.txt"
+    missing = "missing.txt"
+    ignored = "ignored.txt"
+    # Comparison should cover at least .csv and .txt files
+    mismatch = "mismatch.csv"
+    extra = "extra.txt"
+    subfolder = Path("folder")
+    # This file exists in both expected and actual, should not raise any alerts because it contents matches
+    # apart from linebreaks
+    create_folder_and_write_text(expected / subfolder / matching, "Line1\r\nLine2")
+    create_folder_and_write_text(actual / subfolder / matching, "Line1\nLine2")
+    # This file only exists in the expected results, and should create an error

```

Comment:
This comment is a bit confusing because it appears to apply to line 143 (containing 'extra') as well, but in actual fact we do not expect 'extra' to appear in messages

Reply:
I've tried to clarify.